lib/FrontendTool/FrontendTool.cpp

```diff
@@ -493,20 +562,22 @@ bool ABIDependencyEvaluator::isOverlayOfClangModule(ModuleDecl *swiftModule) {
 }
 
 bool ABIDependencyEvaluator::isFakeCycleThroughOverlay(
-    ModuleDecl **sandwichModuleIter) {
-  assert(sandwichModuleIter >= searchStack.begin()
-         && sandwichModuleIter < searchStack.end()
-         && "sandwichModuleIter points to an element in searchStack");
-  // The sandwichedModule must be a Clang module.
-  if (!(*sandwichModuleIter)->isNonSwiftModule())
+    ModuleDecl **startOfCycle) {
+  assert(startOfCycle >= searchStack.begin()
+         && startOfCycle < searchStack.end()
+         && "startOfCycleIter points to an element in searchStack");
+  // The startOfCycle module must be a Clang module.
+  if (!(*startOfCycle)->isNonSwiftModule())
     return false;
-  auto nextModuleIter = sandwichModuleIter + 1;
-  if (nextModuleIter == searchStack.end())
-    return false;
-  // The next module must be a Swift overlay for a Clang module
-  if ((*nextModuleIter)->isNonSwiftModule())
-    return false;
-  return isOverlayOfClangModule(*nextModuleIter);
+  // Next, we must have zero or more modules followed by a Swift overlay for a
+  // Clang module.
+  for (auto moduleIter = startOfCycle + 1;
+       moduleIter != searchStack.end(); ++moduleIter) {
```

Comment:
Nit: Now that we're on the same page about the desired behavior, I realize that the code might be clearer if you used `std::any_of` to express this loop in a more functional style. (By no means mandatory, but please consider it.)